Add bookings

strannicamira opened the pull request.

strannicamira committed:
Add @Entity

strannicamira committed:
Add schema.sql

strannicamira committed:
Fix item owner

strannicamira committed:
Add @GeneratedValue

strannicamira committed:
Refactor ItemDto usage

strannicamira committed:
Refactor UserDto usage

strannicamira committed:
Update application.properties

strannicamira committed:
Fix application.properties

strannicamira committed:
Add Item Repo
Run Ok

strannicamira committed:
Add findByOwnerIdAndId()
Run Ok

strannicamira committed:
Add QItem
Fix List<ItemDto> getUserItems(Integer userId, String text)
Run Ok

strannicamira committed:
Add User Repo
Run Ok

strannicamira committed:
Clean import
Run Ok

strannicamira committed:
Fix Patch User
Run Ok
Run Postman User Folder Ok

strannicamira committed:
Fix Item
Run Ok
Run Postman Item Folder Ok

strannicamira committed:
Update Booking
Not Ok Run
Not Ok Run Postman Booking Folder

strannicamira committed:
Update Booking
Ok Run
Not Ok Run Postman Booking Folder

strannicamira committed:
Update Booking
Ok Run
Not Ok Run Postman Booking Folder:
TODO: Fix Booking 1 create from user 1 to item 2 current and next

strannicamira committed:
Update return to BookingOutDto
Ok Run
Not Ok Run Postman Booking Folder:
Fix Booking 2 get by user 1 (booker)
TODO: Fix Booking get all for user 1

strannicamira committed:
Fix Booking LocalDateTime
Ok Run
Not Ok Run Postman Booking Folder:
TODO: Booking get all for user 1

strannicamira committed:
Update Booking:
Add sort
Ok Run
Almost Ok Run Postman Booking Folder:
TODO: Item 2 get by user 4 (owner) with  bookings

strannicamira committed:
Refactor getUserBookings & getUserItemsBookings
Ok Run
Almost Ok Run Postman Booking Folder:
TODO: Item 2 get by user 4 (owner) with  bookings

strannicamira committed:
Hallelujah!
Fix Item 2 get by user 4 (owner) with  bookings
Item 2 get by user 1 without  bookings
Item 2 get by user 5 without  bookings

Build Run OK
Postman Run Ok/Not Ok
TODO: Item get all user 4 with bookings

strannicamira committed:
Fix Item get all user 4 with bookings

TODO: Item 1 get from user 1 (owner) without comments

Build Run OK
Postman Run Ok/Not Ok

strannicamira committed:
Add Comments
Build Run OK
Postman Run Ok/Not Ok
TODO: Item get all user 4 with bookings

strannicamira committed:
Fix las date booking
Build Run OK
Postman Run Ok/Not Ok
TODO: Item get all user 4 with bookings

strannicamira committed:
Add comments to item
Build Run OK
Postman Run Ok/Not Ok
TODO: Comment from user 4 to item 1 without booking failed

strannicamira committed:
Fix ex
Build Run OK
Postman Run Ok/Not Ok
TODO: Booking get all for user 1 by CURRENT state

strannicamira committed:
Unbelievable!
Fix cur bookings order
Build Run OK
Postman Run Ok

strannicamira committed:
Checkstyle (1)
Clean up and format

strannicamira committed:
Checkstyle (2)
Clean up and format

strannicamira committed:
Checkstyle (3)
Clean up and format

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/booking/BookingService.java:
```diff
@@ -0,0 +1,32 @@
+
+@Transactional(readOnly = true)
+public interface BookingService {
+    @Transactional
```
Эту аннотацию лучше повесить на методы реализации BookingServiceImpl, все же интерфейс - это внешний контракт, а работа с транзакциями - это детали реализации :)
Вернее там аннотация уже есть, то есть тут ее можно просто убрать

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/booking/BookingOutGetDto.java:
```diff
@@ -0,0 +1,28 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class BookingOutGetDto {
```
BookingOutGetDto - дословно "внешнее бронирование получить DTO"

А надо ли иметь столько разных Dto, они же путают, выше пример? Они путают, потом их надо постоянно синхронизировать между собой если базовая сущность будет меняться. Ну и разбираться в таким мнодестве объектов трудновато, есть ли тут возможность обойтись одним лишь BookingDto? :)

strannicamira replied:
Удалила :)

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/item/Item.java:
```diff
@@ -0,0 +1,36 @@
+    @Column(name = "is_available")
+    private Boolean available;
+    @ManyToOne(fetch = FetchType.LAZY)
+//    @ToString.Exclude//TODO:
```
На самом деле все верно, лучше исключить поле из toString() 

Поле отмечено как LAZY, и, соответственно, при обращении к этому полю будет инициирован SQL-запрос. И он может быть инициирован случайно, допустим, когда мы выводим ту же отладочную информацию в лог :)

strannicamira replied:
Да просто забыла раскоментить))

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -1,52 +1,170 @@
+        List<Item> items = repository.findByOwnerId(userId);
+        List<ItemWithBookingDto> dtos = new ArrayList<>();
+        for (Item item : items) {
+            ItemWithBookingDto itemWithBookingDto = getItemWithBooking(userId, item.getId());
```
А не вызывается и тут случаем в цикле кучка SQL-запроосов? Лучше всегда стараться вытащить все необходимые данные за минимальное количество запросов :)

strannicamira replied:
Понимаю. 
Я думала, как лучше, придумала, что, наверное, самое оптимальное переюзать уже имеющийся метод...
Не сомневаюсь, что у меня ещё немало чего можно оптимизировать в плане запросов к бд...
Но пока тут что-то не придумывается по-другому. Оставлю пока так?) Может потом дойдёт...

EugeneLenkevich replied:
Ну оставь, но, если что, тут поможет запрос select ... where id in (...)

Или в виде метода спрингового репозитория

List<Item> findByIdIn(List<Long> ids);

EugeneLenkevich reviewed (changes requested):
Привет, Ирина, спасибо за PR

В целом все отлично :)
Есть совсем небольшие пометки, посмотри комментарии ниже, пожалуйста

strannicamira committed:
Review (1)
Delete BookingOutGetDto and so on
Clean services from Transactional annotation
Back  @ToString.Exclude for Item.owner

EugeneLenkevich reviewed (approved):
Все супер, спасибо!
